vpr/src/route/rr_graph2.cpp

```diff
@@ -970,8 +971,12 @@ static void load_chan_rr_indices(const int max_chan_width,
     }
 }
 
-static void load_block_rr_indices(const DeviceGrid& grid,
+/* As the rr_indices builders modify a local copy of indices, use the local copy in the builder 
```

Comment:
I don't quite understand this.  There is both a local couple of the rr_indices data structure and another copy in the RRGraphBuilder.  It looks like some data is put in the version in the builder (with add_node) and some is in the local copy (with direct data structure access). I'm not sure where the local copy is copied back to the RRGraphBuilder copy.

It would be good to clean this up sooner rather than later.  The circular dependency in the header files (rr_graph2.h etc.) should be resolvable by refactoring the header files / making other headers.  Ideally we'd just use the builder functions when adding nodes to the rr_indices data structure, and would have only one copy.

Reply:
This is indeed confusing. We actually share the data pointers between the ``rr_indices`` and ``RRGraphBuilder`` when initializing the ``DeviceContext``.  So the two data structures are always synchronized.

https://github.com/tangxifan/vtr-verilog-to-routing/blob/4a8443b135bdbd97d8e97da73f33ce2cedd0d7e6/vpr/src/base/vpr_context.h#L166-L180

Currently, the data ownership remains in the ``rr_indices`` because I avoid to modify many codes in rr_graph2.cpp which will make this PR tedious. I just modify a very small part to showcase that the new data structure is working fine. 

And you are right. This code should be cleaned up ASAP. I also stated in the TODO of the comment in LINE 975. I will start massively deployment of the ``RRGraphBuilder`` in the function ``alloc_and_load_rr_indices()`` in next PR.

Reply:
OK.  I think you should comment the pointer aliasing right now though, so it's understandable how it works at this point. I didn't figure out that you'd aliased the pointers (both point at the same storage) and that's why the code worked.